### ISSUE TITLE
HTML Pre-processor CLI

### ISSUE THREAD

**Issue by nc0fr** [nc0fr/web#1]
*(This is a post-mortem for the Woody experiment, now becoming part of NWT)*

The HTML pre-processor is a CLI and an implementation of a DOM-API-like scripting system to allow meta programming over HTML and XHTML documents.

The pre-processing allows reflecting over HTML documents, which can be then used for templating, components replacing, dead link checking, static analysis, &c.

More coming soon

**Comment by nc0fr** [comment]
The overview of the system is the following: given an HTML document and a manipulation script (a script which uses the HTML Document Manipulation API), apply the manipulation on the HTML and returns a new HTML.

For illustration:

Start with the following HTML file `foo.html`.

```html
<!DOCTYPE html>
<html>
<title>Foo</title>
<meta charset="utf-8">
<main>
   <Paragraph>Lorem Ipsum</Paragraph>
</main>
</html>
```
Our goal is to replace our custom `<Paragraph>` component with a `<p>` tag.

We can write the following script (`paragraph.component`) using our HTML Document Manipulation API. (Python for now)

```python
#!python3

import nwt.html_processor as hp


PARAGRAPH_TAG = "Paragraph"
paragraph_component = lambda e: 
  e.replace_markup(f"<p>{e.inner_html}")

document = hp.parse(stdin)
document = document.locate(PARAGRAPH_TAG)
                   .without_attributes()
                   .for_each(paragraph_component)

hp.return(document)
```

We can now process our HTML document using the command line:

```bash
chmod +x ./paragraph.component
mkdir -p build/foo
./paragraph.component <foo.html >build/foo/index.html
```

Which creates `build/foo/index.html`:
```html
<!DOCTYPE html>
<html>
<title>Foo</title>
<meta charset="utf-8">
<main>
   <p>Lorem Ipsum
</main>
</html>
```

The particularity of this system is that it allows the creation of small and pluggable processors/compilers. The modularity is a great advantage.

**Comment by nc0fr** [comment]
I believe this kind of scripting is more natural (at least it is for me) than direct iteration over HTML files as one could do using AWK or Perl.

**Comment by nc0fr** [comment]
The basics of the system is just that out HTML Manipulation API is just a well-designed abstraction over any parser used by the system (presumably Tree-sitter)

**Comment by nc0fr** [comment]
We will use Lua for scripting

**Comment by lilmaax** [comment]
Yeah finally you move to wise choices. Bazel god thanks you for your philosophy.
